### PR TITLE
Shuttle records console to allow copying ship deeds

### PR DESCRIPTION
**Pull request by GreaseMonk**
## About the PR

This adds a new Shuttle records console to the game that allows the station respresentative to put the deed of a ship onto the target ID card.

Copying is only enabled for ships NOT purchased at black market / syndicate

## Why / Balance

According to dvir this has been a long requested feature for over a year and it was time to make it happen.

## How to test

Things to test:
- Anyone can view the UI and its data
- only works if enough balance in the stations bank account (10,000 spesos)
- only works for SR (give yourself all access in the menu)
- only works if there is an id card in the console
- only id cards fit inside the console and only one fits in it and can be taken out again with alt click.
- only works on grids with station bank account component.
- does not work on ships
- does not list syndicate or freelancer purchased shipyard shuttles (so it should not list pirate ship etc)
- lists any shuttle purchased on frontier, nfsd, exped lodge, and also check mothership consoles if you can.
- check that the console is mapped onto frontier station, it is in the SR office.
- ensure that after using the console to set id card deed, a admin log is logged giving enough details.

## Media

image

image

## Requirements

- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes

none

**Changelog**

:cl:
- add: Add shuttle records console to SR office allowing to copy deeds

### CODE REVIEW

**Comment by CautionDev** [comment]
Maybe a remove deed button could also be added? That way SR can also remove deeds for ships that are inoperable and allow the owner to purchase a new one. Also would help in case the SR accidentally adds the incorrect ship and needs to fix it.

**Comment by GreaseMonk** [comment]
> Maybe a remove deed button could also be added? That way SR can also remove deeds for ships that are inoperable and allow the owner to purchase a new one. Also would help in case the SR accidentally adds the incorrect ship and needs to fix it.

it serves as a transaction history now, removing them could be added in a new PR if needed, im trying to keep PR's small

**Comment by GreaseMonk** [comment]
aight it should be ready for review

**Comment by whatston3** [comment]
Would like to see:
* Which ID card (name, job) is currently inserted into the computer, could add an eject button as well a la ID card console.
  * Don't want to accidentally assign yourself a ship as the SR.
* Which ship, if any, is currently registered to the ID card inserted.
* Whether or not the ship was purchased with a voucher.
* Some ability to hide ships that do not currently exist.
  * Further filtering would be nice to have but a bit much to add in this PR
  * All/active should be straightforward and a big QOL improvement for users late in the round.
* Shuttle status should read something to the effect of "Active" vs. "Sold" vs. "Deleted(?)" (not sold, but no corresponding entity - generally acts of the gods)

Ships that were purchased on a voucher should not sell for any meaningful value, even if reassigned - I _do_ think it's useful to be able to reassign derelict voucher ships, if only to reduce the number of floating hulks, so reassignment and salvage/sale at no value would be handy.

I'll see if I can get an MVP of this.

This system's useful, seems like a good addition, should be handy.  Good stuff.

**Comment by whatston3** [comment]
Righto, got this up: [whatston3:2024-10-25-sr-id-card-recovery-suggestions](https://github.com/new-frontiers-14/frontier-station-14/compare/fs-14-sr-id-card-recovery...whatston3:2024-10-25-sr-id-card-recovery-suggestions?expand=1)

This uses a dictionary to store records by entity UID (no dupes) and adds functions for record lookup and record updates to the ShuttleRecordsSystem.

To cut down on bandwidth, I made the record set in the BUI updates nullable - a null record set implies no change in the record set, and the UI falls back on the last received set of records.

The shuttle records console UI currently updates on ship purchase, and it _should_ update on ship sale, but **isn't currently doing that** (not sure why, might need to come back to this).  Changes to ship names are now reflected on the record console.

Voucher abuse was removed, and added as a flag to the ShuttleDeed.  Since we also aren't guaranteed that a voucher sale is the only source of sale of voucher ships, I added a check to the purchase to handle that.

Give it a look to make sure it's up to your standards, but behaviourally, this seems like an improvement to me.

UI currently looks like this on that branch (layout needs work, this is a bit slapdash (just a bit :^) ), though I think the new fields make sense).
image

**Comment by GreaseMonk** [comment]
In general, please please dont take this as an offence, If you find things that should be in it- I can do it, you don't have to do it for me uwu. That being said, we should talk about all the new features you implemented. I do like some of the additions but all of them are open for discussion.

Lets do any new features in a new PR because I meant for this to be as simple as possible.

I would like to hear your thoughts whether you agree/disagree with my comments:

> This uses a dictionary to store records by entity UID (no dupes)
I wanted this system to be readable, so in order to do so, i would prefer to have a check if it is already stored before adding it, rather than a key value pair, and this is primarily because you can't name dictionary key and value params. Making a dictionary, in my opinion- i would rather make a new class that holds these variables (like how it is now) so we can name them so you can immediately know what the EntityUid is without having to comment it everywhere.

> To cut down on bandwidth, I made the record set in the BUI updates nullable
This overcomplicates a feature that is only used by one client at a time, don't you think the bandwidth is negligable and overkill ? Gonna look further into this  and going to decide wether to use it depending on how readable it looks and based on your opinion!

> Shuttle status should read something to the effect of "Active" vs. "Sold" vs. "Deleted(?)" 
I kept it simple because I didn't want to give away any information about its whereabouts since this is simply a ship sales transaction log.

From what i gathered some things that should be done still is:
- Ships purchased with a voucher can't be copied.
- Indicate if a ship was purchased using a voucher.
- ID card details and eject button
- I noticed some bug with the list items disabled when switching still needs a fix

**Comment by whatston3** [comment]
>This overcomplicates a feature that is only used by one client at a time, don't you think the bandwidth is negligable and overkill ? Gonna look further into this and going to decide wether to use it depending on how readable it looks and based on your opinion!

Sending 100+ records, each with multiple strings, per update by end of round, _for actions that do not change that set_ seems silly.
Straightforward logic to keep old records around (admittedly still duped in the BUI, but that's a small hit to client-side memory)
This was one of the issues I brought up with Quigley's PR - "you're going to get a lot of info on each ship, you might need to paginate these"
I could see this being useful for the NFSD, so I don't think the argument of "it's only one" will last long - a view on this data would be useful, at minimum, for the STC and detective.

> Lets do any new features in a new PR
I would argue this isn't a new feature, it's all changes to the ShuttleRecords, the things that populate them, and the console that views them.  At the moment, that is the system in its entirety.

> I wanted this system to be readable, so in order to do so, i would prefer to have a check if it is already stored before adding it
If you have multiple entities with the same UID, that is an issue.
Lookup in larger tables should be logarithmic at worst.
Again, in the first PR, I said that indexing would be a pain in the ass.  This is the simplest index.  More may be useful.

>you can't name dictionary key and value params
But you can name the dictionary, and something like "VesselsByNetEntity" is pretty unambiguous with the key being a NetEntity and the value being something along the lines of VesselRecord.

> I kept it simple because I didn't want to give away any information about its whereabouts since this is simply a ship sales transaction
Fair enough, but "no updates in a while" doesn't imply "this no longer exists" - a derelict ship out drifting at 42 km+ is still "active in sector", but knowing that "so and so sold a ship", which _you receive a radio callout for_, is reasonable to know - if you wanted to keep that ambiguous if they sold it at LPBravo, great, cache the value and don't rely on ad hoc entity lookups

> I do like some of the additions but all of them are open for discussion.
Yes, it is a suggestion, and not pushed into the PR.

**Review by whatston3 — DISMISSED** [review]
Left a request for the following improvements.  I think these are minimal for a reasonable UX.

https://github.com/new-frontiers-14/frontier-station-14/pull/2214#issuecomment-2436485408

Transferring ownership of derelict vessels acquired through vouchers is reasonable, and allows players to maintain the level of garbage on the server.

File: Content.Client/_NF/ShuttleRecords/UI/ShuttleRecordsWindow.xaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+﻿<controls:FancyWindow xmlns="https://spacestation14.io"
+                      xmlns:graphics="clr-namespace:Robust.Client.Graphics;assembly=Robust.Client"
+                      xmlns:controls="clr-namespace:Content.Client.UserInterface.Controls"
+                      Title="{Loc 'shuttle-records-console-window-title'}"
+                      MinSize="700 300">
```
Comment (whatston3) [inline Content.Client/_NF/ShuttleRecords/UI/ShuttleRecordsWindow.xaml:5]: If you increase the minimum size of the window, you avoid weird issues with the shuttle records window overlapping the top panel.
```suggestion
                      MinSize="700 510">
```

File: Content.Client/_NF/ShuttleRecords/UI/ShuttleRecordsWindow.xaml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                <BoxContainer Orientation="Vertical" Margin="0 0 0 5">
+                    <Label Text="{Loc 'shuttle-records-search'}" StyleClasses="LabelSubText" Margin="0 0 5 0" />
+                    <BoxContainer Orientation="Horizontal">
+                        <LineEdit Name="SearchText" Access="Public" MinHeight="30" SetWidth="200" />
+                        <Button Name="ClearSearchButton" Access="Public" Text="✖" Margin="5 0 0 0" />
+                    </BoxContainer>
+                    <CheckBox Name="ShowActiveOnlyCheckbox" Text="{Loc 'shuttle-records-show-active-only'}" />
+                </BoxContainer>
+
+                <BoxContainer Orientation="Vertical" MinSize="250 300">
```
Comment (whatston3) [inline Content.Client/_NF/ShuttleRecords/UI/ShuttleRecordsWindow.xaml:35]: If you add two VerticalExpand values (everything cascading down to the shuttle list), you end up getting additional shuttle records when expanding the window (good, IMO)
```suggestion
            <BoxContainer Orientation="Vertical" Margin="0 0 10 0" VerticalExpand="True">
                <BoxContainer Orientation="Vertical" Margin="0 0 0 5">
                    <Label Text="{Loc 'shuttle-records-search'}" StyleClasses="LabelSubText" Margin="0 0 5 0" />
                    <BoxContainer Orientation="Horizontal">
                        <LineEdit Name="SearchText" Access="Public" MinHeight="30" SetWidth="200" />
                        <Button Name="ClearSearchButton" Access="Public" Text="✖" Margin="5 0 0 0" />
                    </BoxContainer>
                    <CheckBox Name="ShowActiveOnlyCheckbox" Text="{Loc 'shuttle-records-show-active-only'}" />
                </BoxContainer>

                <BoxContainer Orientation="Vertical" MinSize="250 300" VerticalExpand="True">
```

Output at default size (100% scaling):
image

Extra records when grown:
image

File: Resources/Prototypes/_NF/Entities/Structures/Machines/Computers/computers_shuttle_records.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
+﻿- type: entity
+  name: shuttle records console
+  parent: [BaseStructureDisableToolUse, BaseStructureIndestructible, BaseComputer]
+  id: ComputerShuttleRecordsBase
+  description: Used to recover lost ID cards
```
Comment (whatston3) [inline Resources/Prototypes/_NF/Entities/Structures/Machines/Computers/computers_shuttle_records.yml:5]: Marginally more descriptive?
```suggestion
  description: Used to browse shuttle records and recover or reassign shuttle ownership to IDs.
```

**Review by whatston3 — COMMENTED** [review]
Left a few comments, but this looks very good!

Might recommend putting the ship class in the info panel to the right, if only for the sake of late join SRs - for example, what ship type is this?
image

One small side effect to note - when selling a ship, when the state is updated, the entity is still believed to be around, but only on that update, any other change resolves properly.  Really not a huge deal, and if the ship's active status was put into the state, easy enough.  Not a big deal now, but if we supported multiple status types (e.g. impounded?), might make sense.

Impressive polish, wish I had such a knack with UI.  Perhaps in time.  Nice feature, hope the SR gets some better QOL.

**Review by whatston3 — APPROVED** [review]
Thanks for the feature, pardon the requests.  Happy with the state of this.

As a note for later, could think about putting it on the Bottleneck.

**Review by whatston3 — APPROVED** [review]
Newest changes look good, approving with latest changeset.

Looking forward to this feature, should be nice QoL for ship reassignments, and a nice thing to do for the SR.

**Pull request merged**